src/shared/sol-util.c

```diff
@@ -52,6 +81,104 @@ sol_util_memdup(const void *data, size_t len)
     return ptr;
 }
 
+double
+sol_util_strtodn(const char *nptr, char **endptr, size_t len, bool use_locale)
+{
+    char *tmpbuf, *tmpbuf_endptr;
+    double value;
+
+#if defined(HAVE_LOCALE) && defined(HAVE_STRTOD_L)
+    if (!use_locale) {
+        if (!init_c_locale()) {
+            /* not great, but a best effort to convert something */
+            use_locale = false;
+            SOL_WRN("could not create locale 'C', use current locale.");
+        }
+    }
+#endif
+
+    /* NOTE: Using a copy to ensure trailing \0 and strtod() so we
+     * properly parse numbers with large precision.
+     *
+     * Since parsing it is complex (ie:
+     * http://www.netlib.org/fp/dtoa.c), we take the short path to
+     * call libc.
+     */
+    tmpbuf = strndupa(nptr, len);
+
+    errno = 0;
+#ifdef HAVE_LOCALE
+    if (!use_locale) {
+#ifdef HAVE_STRTOD_L
+        value = strtod_l(tmpbuf, &tmpbuf_endptr, c_locale);
```

Comment:
It's not that simple or useful to just emulate `strtod_l()` in general.  In this patch we mimic a well known locale, "C", and take short paths. If we say we support users giving any `locale_t`, not only "C", then we must do much more work.